### PR TITLE
Optimizations in phy.plot

### PR DESCRIPTION
**Pull request by rossant**

### CODE REVIEW

**Comment by yger** [comment]
I can see some potential problems with the PR. Minor, but still

- Toggling show labels has no effect, both in Waveform view and in TraceView (in Waveform view, in fact, it changes the zoom)

- Sometimes, I can find templates that have a clear shape, with overlapping waveforms, but all channels are grey. How is the blue gradient computed? It seems that there is something wrong here, as, for every template, the chanell with maximal amplitude should be blue, am I right? Or can it be than small templates, even if well isolated, can be gray because the gradient is something computed globally, over all templates. I'm illustrating that with a snapshot taken from a small dataset

gradient

- For large dataset with more than 1000 channels, there is an offset between channels and numbers (see attached) So I suspect the offset is not properly computed somewhere
large

- This offset is such that if the "Zoom on channel" option is activated, it will center the screen on the numbers and not the channels, that are far apart. Just to mention

- With the large dataset, I do not understand how the View is centered on the template. The zoom level is kept constant when you change template, which is great, but could it be centered on the channel where the template has its minimal value? If this is the case, then this is not working. 

- I know I do not have much to propose yet, I will think about it, but again, here is the crowding issue I'm talking about with the TraceView and numerous channels. Two problems here: one is that this is hardly readible, and the shaded colors are great, but to see them, you need to overlap traces. So the key point is to restrict the Traces to a subset of channels. 

trace

Thanks !

Pierre

**Comment by nsteinme** [comment]
Hey, I have tested it now too. 

I can confirm at least one of Pierre's observations - the coloration of the waveforms isn't working properly (for me it seems that the largest channel is now always grey). It also seems like traceview colors all got dimmer somehow? 

I don't see the "numbers offset from waveforms" (for me they look fine) and "auto-zoom doesn't center on the channel" (for me it does, although it does zoom out in the x-dimension every time it does this). 

So basically everything seems to be working for me! But I agree the speed-ups seem small, sorry :( I still get about 2-3 seconds between a click and visualization, I would estimate, with little difference whether it is the first time I've clicked on that cluster or the second.

**Comment by rossant** [comment]
Thanks for the feedback! I just pushed a fix in phy@plot-optim2, could you try again? It should be a bit faster.

I'll get back to all the other comments later. In the meantime, you should know that the next major optimization I'm going to do is a mode where you just unmasked channels are shown in the waveform and trace view. This should make the GUI a lot faster.

**Comment by rossant** [comment]
@nsteinme FYI, with the latest fix, switching from one cluster to the next takes 0.5-1 second on my laptop (once they are in the cache). But I agree it is too slow, still, and the "unmasked" optimization should make it faster.

**Comment by yger** [comment]
Ok, great, Ill try it asap. Looking forward for the unmasked option, but I think the general responsiveness of the GUI starts to be good. Let me know for the others minors bugs that I spotted if I'm the only one seeing them...

**Comment by nsteinme** [comment]
Hey Cyrille, fantastic, yes it is faster in the latest fix! It's pretty
darn close now to "as fast as I would be able to use".

There seems to be at least one new bug: when I make a merge, the cluster I
merged with doesn't disappear from similarity view?

Also, I keep seeing this:
14:20:37 [I] gui_component:184    Merge clusters 7, 68 to 511.
14:20:38 [E] cli:37               An error has occurred (KeyError): 511
14:20:39 [E] cli:37               An error has occurred (KeyError): 511
14:20:42 [E] cli:37               An error has occurred (KeyError): 511

When I do various selections - I think it might be coming from amplitude
view which doesn't seem to always update properly? This was happening in
the latest master branch too (I think).

If I try to display the template for a merged cluster, the waveform view
becomes empty and this shows in the command prompt:
14:23:44 [E] cli:37               An error has occurred (IndexError): index
0 is
 out of bounds for axis 0 with size 0

Also this thing is still happening in the feature view?

[image: Inline image 1]

i.e. black boxes around some (but not all?) of the points.

**Comment by rossant** [comment]
* A few bugs have been fixed.
* As requested, the color of the non-selected spikes are dimmer, so that selected spikes are more clearly visible.
* As I said, the next thing I'll do is the unmask mode.
* I cannot reproduce the bugs related to Pierre's large datasets - can you send me the dataset? Idem for the unmask color in the waveform view.

**Comment by yger** [comment]
I can send you the dataset, but it may be faster if you could pop up at the lab one day. As you want. Otherwise tomorrow I will try to upload it and send you a link. With latest update, I agree this is faster, but I still have the gray color sometimes all over templates in the waveform View.

**Comment by nsteinme** [comment]
Hi Cyrille, it is messing up the unmask color on every cluster for me in multiple datasets including the one I've sent you (20151102). You sure it's not doing it for you? Looks like this:

image


You see how there's one big one that's incorrectly grey there, right? Also if I have the waveform view set like that then go to the next cluster (down arrow) it changes the zoom to look like this, if I have auto-zoom on: 

image


I used to just turn off auto-zoom but now that I can switch clusters so fast it's back to being a potentially very useful feature :) It needs to keep the same frame but just translate it in x/y. 

Also I disagree about the dimming of traceview. The main thing I look for there is the residual trace and I can't see it at all. Maybe my eyes are getting old. Can it be optional at least? 

Thanks -

**Comment by rossant** [comment]
> Also I disagree about the dimming of traceview. The main thing I look for there is the residual trace and I can't see it at all. Maybe my eyes are getting old. Can it be optional at least?

Wait, what I changed was the color of the non-selected spikes. Are you talking about these? Or the traces themselves?

**Comment by rossant** [comment]
The mask bug is fixed :) (due to the GPU optimizations I have to multiply the masks by 0.999)

**Comment by rossant** [comment]
@nsteinme the auto-zoom feature in the waveform view should work

@yger I think that several of the bugs you mention are actually one single bug that is only seen on your large dataset. I can't pass by at IdV right now, so it would be great if you could give me access to your dataset. Any way I can scp to one of your servers?

**Comment by nsteinme** [comment]
It appeared to me that the non-highlighted parts of the traces, and the "residual" parts of the traces, were dimmer than before. 

Here is before: 
image

Here is what I see now:

image

I think it seems clear that the background trace color, and the color of the "residual" traces, has gotten dimmer? In any case I vote that it is too dim :)

**Comment by rossant** [comment]
whoops this is unintended... let me fix it

**Comment by rossant** [comment]
OK it should be better now, let me know if it isn't the case

**Comment by yger** [comment]
I'll try to send you test files this afternoon, you can not scp the lab, unfortunatly.... more than thousands channels, dummy recording, but still, this will give you a taste of what the GUI should handle

Best

Pierre
